File: services/inventory-service/src/routes/health.ts
import { Router, Request, Response } from "express";
import { ping, timescaledbVersion } from "../config/database";
import { logger } from "../config/logger";

export const healthRouter = Router();

// ---------------------------------------------------------------------------
// GET /health
// ---------------------------------------------------------------------------
// Returns HTTP 200 when the service and its dependencies are healthy,
// HTTP 503 when any required dependency is unreachable.
//
// Response shape:
// {
//   "status":  "ok" | "degraded",
//   "uptime":  123.45,                 // process uptime in seconds
//   "timestamp": "2025-01-01T00:00:00Z",
//   "checks": {
//     "database": {
//       "status":      "ok" | "error",
//       "latencyMs":   4,
//       "timescaledb": "2.14.2"        // null when TimescaleDB is not installed
//     }
//   }
// }
// ---------------------------------------------------------------------------

interface DatabaseCheck {
  status:      "ok" | "error";
  latencyMs?:  number;
  timescaledb: string | null;
  error?:      string;
}

interface HealthBody {
  status:    "ok" | "degraded";
  uptime:    number;
  timestamp: string;
  checks: {
    database: DatabaseCheck;
  };
}

healthRouter.get("/health", async (_req: Request, res: Response) => {
  let dbCheck: DatabaseCheck;

  try {
    const [latencyMs, tsdbVersion] = await Promise.all([
      ping(),
      timescaledbVersion(),
    ]);

    dbCheck = {
      status:      "ok",
      latencyMs,
      timescaledb: tsdbVersion,
    };
  } catch (err) {
    const message = err instanceof Error ? err.message : String(err);
    logger.warn("Health check: database unreachable", { error: message });

    dbCheck = {
      status:      "error",
      timescaledb: null,
      error:       message,
    };
  }

  const healthy = dbCheck.status === "ok";

  const body: HealthBody = {
    status:    healthy ? "ok" : "degraded",
    uptime:    Math.round(process.uptime() * 100) / 100,
    timestamp: new Date().toISOString(),
    checks: {
      database: dbCheck,
    },
  };

  res.status(healthy ? 200 : 503).json(body);
});

// ---------------------------------------------------------------------------
// GET /health/live
// ---------------------------------------------------------------------------
// Lightweight liveness probe for Kubernetes — does not check dependencies.
// Returns 200 as long as the Node process is running.
// ---------------------------------------------------------------------------

healthRouter.get("/health/live", (_req: Request, res: Response) => {
  res.status(200).json({ status: "ok" });
});

// ---------------------------------------------------------------------------
// GET /health/ready
// ---------------------------------------------------------------------------
// Readiness probe — same as /health but semantically signals that the service
// is ready to receive traffic (load balancer / k8s readiness gate).
// ---------------------------------------------------------------------------

healthRouter.get("/health/ready", async (_req: Request, res: Response) => {
  try {
    await ping();
    res.status(200).json({ status: "ok" });
  } catch {
    res.status(503).json({ status: "not_ready", reason: "database_unreachable" });
  }
});

File: services/moisture-service/src/routes/health.py
"""
Health check endpoints.

GET /health        — full dependency check; 200 ok / 503 degraded
GET /health/live   — liveness probe (no dependency checks)
GET /health/ready  — readiness probe (database only)
"""

from __future__ import annotations

import time
from typing import Literal

from fastapi import APIRouter
from pydantic import BaseModel

from src.db import ping
from src.logger import get_logger

log = get_logger(__name__)
router = APIRouter(tags=["health"])

_start_time = time.time()


# ---------------------------------------------------------------------------
# Response models
# ---------------------------------------------------------------------------

class DatabaseCheck(BaseModel):
    status: Literal["ok", "error"]
    latency_ms: int | None = None
    timescaledb: str | None = None
    error: str | None = None


class ModelCheck(BaseModel):
    status: Literal["ok", "not_loaded"]
    device: str | None = None


class HealthChecks(BaseModel):
    database: DatabaseCheck
    model: ModelCheck


class HealthResponse(BaseModel):
    status: Literal["ok", "degraded"]
    uptime_seconds: float
    timestamp: str
    checks: HealthChecks


class LiveResponse(BaseModel):
    status: Literal["ok"]


class ReadyResponse(BaseModel):
    status: Literal["ok", "not_ready"]
    reason: str | None = None


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------

def _uptime() -> float:
    return round(time.time() - _start_time, 2)


def _now_iso() -> str:
    from datetime import datetime, timezone
    return datetime.now(tz=timezone.utc).isoformat()


def _model_check() -> ModelCheck:
    """
    Report whether the PyTorch model is loaded.
    The model registry is imported lazily so this file has no hard
    dependency on torch at import time.
    """
    try:
        from src.model import model_registry  # noqa: PLC0415
        if model_registry.is_loaded():
            return ModelCheck(status="ok", device=model_registry.device)
        return ModelCheck(status="not_loaded")
    except ImportError:
        return ModelCheck(status="not_loaded")


# ---------------------------------------------------------------------------
# Routes
# ---------------------------------------------------------------------------

@router.get(
    "/health",
    response_model=HealthResponse,
    summary="Full health check",
    description=(
        "Returns HTTP 200 when all dependencies are healthy, "
        "HTTP 503 when any required dependency is unreachable."
    ),
)
async def health() -> HealthResponse:
    db_ok = False
    db_check: DatabaseCheck

    try:
        latency_ms, tsdb_version = await ping()
        db_check = DatabaseCheck(
            status="ok",
            latency_ms=latency_ms,
            timescaledb=tsdb_version,
        )
        db_ok = True
    except Exception as exc:  # noqa: BLE001
        msg = str(exc)
        log.warning("Health check: database unreachable", error=msg)
        db_check = DatabaseCheck(status="error", error=msg)

    model_check = _model_check()
    healthy = db_ok

    from fastapi.responses import JSONResponse  # noqa: PLC0415
    body = HealthResponse(
        status="ok" if healthy else "degraded",
        uptime_seconds=_uptime(),
        timestamp=_now_iso(),
        checks=HealthChecks(database=db_check, model=model_check),
    )

    if not healthy:
        # FastAPI will serialise the model; return 503 manually
        from fastapi import Response  # noqa: PLC0415
        return JSONResponse(status_code=503, content=body.model_dump())  # type: ignore[return-value]

    return body


@router.get(
    "/health/live",
    response_model=LiveResponse,
    summary="Liveness probe",
    description="Lightweight probe for Kubernetes — no dependency checks.",
)
async def liveness() -> LiveResponse:
    return LiveResponse(status="ok")


@router.get(
    "/health/ready",
    response_model=ReadyResponse,
    summary="Readiness probe",
    description="Signals that the service is ready to receive traffic.",
)
async def readiness() -> ReadyResponse:
    try:
        await ping()
        return ReadyResponse(status="ok")
    except Exception as exc:  # noqa: BLE001
        log.warning("Readiness check failed", error=str(exc))
        from fastapi.responses import JSONResponse  # noqa: PLC0415
        body = ReadyResponse(status="not_ready", reason="database_unreachable")
        return JSONResponse(status_code=503, content=body.model_dump())  # type: ignore[return-value]
